From 51f68d29af871dd961d3f938a858f235b544fa40 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 21 Jun 2024 15:07:04 +0200
Subject: [PATCH 0934/1695] drivers: stm32_rifsc: use only one argument in
 firewall queries

Update driver to adapt to 1 argument in firewall queries.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9b56e063b1607060c1997237509ed9fdaf30c3a8 drivers: stm32_rifsc: use only one argument in firewall queries]
Change-Id: I4e8fc294e372095f22b6735601b64d7f151bd201
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/389469
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 56 +++++++++++++++++------------
 1 file changed, 33 insertions(+), 23 deletions(-)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 8e4569d81..da61ca7f8 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -478,6 +478,7 @@ static TEE_Result stm32_rifsc_check_access(struct firewall_query *firewall)
 	uintptr_t rifsc_base = rifsc_pdata.base;
 	unsigned int cid_reg_offset = 0;
 	unsigned int periph_offset = 0;
+	unsigned int resource_id = 0;
 	uint32_t cid_to_check = 0;
 	unsigned int reg_id = 0;
 	bool priv_check = true;
@@ -488,30 +489,28 @@ static TEE_Result stm32_rifsc_check_access(struct firewall_query *firewall)
 
 	assert(rifsc_base);
 
-	if (!firewall || firewall->arg_count != 2)
+	if (!firewall || firewall->arg_count != 1)
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	/*
 	 * Peripheral configuration, we assume the configuration is as
 	 * follows:
-	 * firewall->args[0]: Firewall ID
-	 * firewall->args[1]: RIF configuration to check, reusing RIFPROT()
-	 *		      macro. Only secure, privilege and static CID
-	 *		      fields are used
+	 * firewall->args[0]: RIF configuration to check
 	 */
-	if (firewall->args[0] >= RIMU_ID_OFFSET)
+	resource_id = firewall->args[0] & RIF_PER_ID_MASK;
+	if (resource_id >= RIMU_ID_OFFSET)
 		return TEE_SUCCESS;
 
-	reg_id = firewall->args[0] / _PERIPH_IDS_PER_REG;
-	periph_offset = firewall->args[0] % _PERIPH_IDS_PER_REG;
-	cid_reg_offset = _OFFSET_PERX_CIDCFGR * firewall->args[0];
+	reg_id = resource_id / _PERIPH_IDS_PER_REG;
+	periph_offset = resource_id % _PERIPH_IDS_PER_REG;
+	cid_reg_offset = _OFFSET_PERX_CIDCFGR * resource_id;
 	cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
 			    cid_reg_offset);
 	seccfgr = io_read32(rifsc_base + _RIFSC_RISC_SECCFGR0 + 0x4 * reg_id);
 	privcfgr = io_read32(rifsc_base + _RIFSC_RISC_PRIVCFGR0 + 0x4 * reg_id);
-	sec_check = (BIT(RIF_SEC_SHIFT) & firewall->args[1]) != 0;
-	priv_check = (BIT(RIF_PRIV_SHIFT) & firewall->args[1]) != 0;
-	cid_to_check = (firewall->args[1] & RIF_SCID_MASK) >> RIF_SCID_SHIFT;
+	sec_check = (BIT(RIF_SEC_SHIFT) & firewall->args[0]) != 0;
+	priv_check = (BIT(RIF_PRIV_SHIFT) & firewall->args[0]) != 0;
+	cid_to_check = (firewall->args[0] & RIF_SCID_MASK) >> RIF_SCID_SHIFT;
 
 	if (!sec_check && seccfgr & BIT(periph_offset))
 		return TEE_ERROR_ACCESS_DENIED;
@@ -535,6 +534,7 @@ static TEE_Result stm32_rifsc_acquire_access(struct firewall_query *firewall)
 {
 	uintptr_t rifsc_base = rifsc_pdata.base;
 	unsigned int cid_reg_offset = 0;
+	unsigned int resource_id = 0;
 	uint32_t cidcfgr = 0;
 
 	assert(rifsc_base);
@@ -542,19 +542,19 @@ static TEE_Result stm32_rifsc_acquire_access(struct firewall_query *firewall)
 	if (!firewall || !firewall->arg_count)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	if (firewall->args[0] >= RIMU_ID_OFFSET)
+	/*
+	 * Peripheral configuration, we assume the configuration is as
+	 * follows:
+	 * firewall->args[0]: Firewall ID of the resource to acquire
+	 */
+	resource_id = firewall->args[0] & RIF_PER_ID_MASK;
+	if (resource_id >= RIMU_ID_OFFSET)
 		return TEE_SUCCESS;
 
-	cid_reg_offset = _OFFSET_PERX_CIDCFGR * firewall->args[0];
+	cid_reg_offset = _OFFSET_PERX_CIDCFGR * resource_id;
 	cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
 			    cid_reg_offset);
 
-	FMSG("CID filtering %s ,static CID%u ,semaphore list:%#x for periph %u",
-	     (cidcfgr & _CIDCFGR_CFEN) ? "enabled" : "disabled",
-	     (cidcfgr & RIFSC_RISC_SCID_MASK) >> SCID_SHIFT,
-	     (cidcfgr & RIFSC_RISC_SEML_MASK) >> SEMWL_SHIFT,
-	     firewall->args[0]);
-
 	/* Only check CID attributes */
 	if (!(cidcfgr & _CIDCFGR_CFEN))
 		return TEE_SUCCESS;
@@ -588,11 +588,16 @@ static TEE_Result stm32_rifsc_set_config(struct firewall_query *firewall)
 	if (res)
 		return res;
 
-	if (!firewall || firewall->arg_count != 2)
+	if (!firewall || firewall->arg_count != 1)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	id = firewall->args[0];
-	conf = firewall->args[1];
+	/*
+	 * Peripheral configuration, we assume the configuration is as
+	 * follows:
+	 * firewall->args[0]: RIF configuration to set
+	 */
+	id = firewall->args[0] & RIF_PER_ID_MASK;
+	conf = firewall->args[0];
 
 	if (id < RIMU_ID_OFFSET) {
 		struct risup_cfg risup = { };
@@ -613,6 +618,11 @@ static TEE_Result stm32_rifsc_set_config(struct firewall_query *firewall)
 				return TEE_ERROR_BAD_PARAMETERS;
 		}
 
+		DMSG("Setting config for peripheral: %u, %s, %s, cid attr: %#"PRIx32", %s",
+		     id, risup.sec ? "Secure" : "Non secure",
+		     risup.priv ? "Privileged" : "Non privileged",
+		     risup.cid_attr, risup.lock ? "Locked" : "Unlocked");
+
 		return stm32_risup_cfg(&rifsc_pdata, &risup);
 	}
 
-- 
2.47.2

